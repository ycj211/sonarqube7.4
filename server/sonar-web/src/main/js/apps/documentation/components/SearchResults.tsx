/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import lunr, { LunrIndex } from 'lunr';
import { sortBy } from 'lodash';
import SearchResultEntry, { SearchResult } from './SearchResultEntry';
import { DocumentationEntry, getUrlsList, DocsNavigationItem } from '../utils';

interface Props {
  navigation: DocsNavigationItem[];
  pages: DocumentationEntry[];
  query: string;
  splat: string;
}

export default class SearchResults extends React.PureComponent<Props> {
  index: LunrIndex;

  constructor(props: Props) {
    super(props);
    this.index = lunr(function() {
      this.ref('relativeName');
      this.field('title', { boost: 10 });
      this.field('text');

      this.metadataWhitelist = ['position'];

      props.pages
        .filter(page => getUrlsList(props.navigation).includes(page.url))
        .forEach(page => this.add(page));
    });
  }

  render() {
    const { query } = this.props;
    const results = this.index
      .search(`${query}~1 ${query}*`)
      .map(match => {
        const page = this.props.pages.find(page => page.relativeName === match.ref);
        const highlights: { [field: string]: [number, number][] } = {};
        let longestTerm = '';

        // remember the longest term that matches the query *exactly*
        Object.keys(match.matchData.metadata).forEach(term => {
          if (
            query.toLowerCase().includes(term.toLowerCase()) &&
            longestTerm.length < term.length
          ) {
            longestTerm = term;
          }

          Object.keys(match.matchData.metadata[term]).forEach(fieldName => {
            const { position: positions } = match.matchData.metadata[term][fieldName];
            highlights[fieldName] = [...(highlights[fieldName] || []), ...positions];
          });
        });

        return { page, highlights, longestTerm };
      })
      .filter(result => result.page) as SearchResult[];

    // re-order results by the length of the longest matched term
    // the longer term is the more chances the result is more relevant
    const sortedResults = sortBy(results, result => -result.longestTerm.length);

    return (
      <>
        {sortedResults.map(result => (
          <SearchResultEntry
            active={result.page.relativeName === this.props.splat}
            key={result.page.relativeName}
            result={result}
          />
        ))}
      </>
    );
  }
}
