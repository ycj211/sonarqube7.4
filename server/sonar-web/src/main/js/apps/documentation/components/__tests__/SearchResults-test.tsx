/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import lunr from 'lunr';
import SearchResults from '../SearchResults';

jest.mock('lunr', () => ({
  default: jest.fn(() => ({
    search: jest.fn(() => [
      {
        ref: 'lorem/origin',
        matchData: {
          metadata: { from: { title: { position: [[19, 5]] }, text: { position: [[121, 4]] } } }
        }
      },
      { ref: 'foobar', matchData: { metadata: { from: { title: { position: [[23, 4]] } } } } }
    ])
  }))
}));

function createPage(title: string, relativeName: string, text = '') {
  return { relativeName, url: '/' + relativeName, title, navTitle: undefined, text, content: text };
}

const pages = [
  createPage(
    'Lorem Ipsum',
    'lorem/index',
    "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book."
  ),
  createPage(
    'Where does it come from?',
    'lorem/origin',
    'Contrary to popular belief, Lorem Ipsum is not simply random text. It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years old. Richard McClintock, a Latin professor at Hampden-Sydney College in Virginia, looked up one of the more obscure Latin words.'
  ),
  createPage(
    'Where does Foobar come from?',
    'foobar',
    'Foobar is a universal variable understood to represent whatever is being discussed.'
  )
];

it('should search', () => {
  const wrapper = shallow(
    <SearchResults
      navigation={['lorem/index', 'lorem/origin', 'foobar']}
      pages={pages}
      query="from"
      splat="foobar"
    />
  );
  expect(wrapper).toMatchSnapshot();
  expect(lunr).toBeCalled();
  expect((wrapper.instance() as SearchResults).index.search).toBeCalledWith('from~1 from*');
});
