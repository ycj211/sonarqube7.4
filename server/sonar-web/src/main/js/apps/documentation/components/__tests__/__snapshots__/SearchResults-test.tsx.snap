// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should search 1`] = `
<Fragment>
  <SearchResultEntry
    active={false}
    key="lorem/origin"
    result={
      Object {
        "highlights": Object {
          "text": Array [
            Array [
              121,
              4,
            ],
          ],
          "title": Array [
            Array [
              19,
              5,
            ],
          ],
        },
        "longestTerm": "from",
        "page": Object {
          "content": "Contrary to popular belief, Lorem Ipsum is not simply random text. It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years old. Richard McClintock, a Latin professor at Hampden-Sydney College in Virginia, looked up one of the more obscure Latin words.",
          "navTitle": undefined,
          "relativeName": "lorem/origin",
          "text": "Contrary to popular belief, Lorem Ipsum is not simply random text. It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years old. Richard McClintock, a Latin professor at Hampden-Sydney College in Virginia, looked up one of the more obscure Latin words.",
          "title": "Where does it come from?",
          "url": "/lorem/origin",
        },
      }
    }
  />
  <SearchResultEntry
    active={true}
    key="foobar"
    result={
      Object {
        "highlights": Object {
          "title": Array [
            Array [
              23,
              4,
            ],
          ],
        },
        "longestTerm": "from",
        "page": Object {
          "content": "Foobar is a universal variable understood to represent whatever is being discussed.",
          "navTitle": undefined,
          "relativeName": "foobar",
          "text": "Foobar is a universal variable understood to represent whatever is being discussed.",
          "title": "Where does Foobar come from?",
          "url": "/foobar",
        },
      }
    }
  />
</Fragment>
`;
